README
Service-WorkHours

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Service::WorkHours

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Service-WorkHours

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Service-WorkHours

    CPAN Ratings
        http://cpanratings.perl.org/d/Service-WorkHours

    Search CPAN
        http://search.cpan.org/dist/Service-WorkHours/


LICENSE AND COPYRIGHT

Copyright (C) 2017 Vincent Tavernier

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Service-WorkHours

0.02    Sun Jul 23 14:03:15 2017 +0200
        Refactored code.
        Added tests.
        Added support for includes.


0.01    Thu Jun 22 21:17:19 2017 +0200
        Initial version.


Makefile.PL
use 5.014;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
	NAME             => 'Service::WorkHours',
	AUTHOR           => q{Vincent Tavernier <vince.tavernier@gmail.com>},
	VERSION_FROM     => 'lib/Service/WorkHours.pm',
	ABSTRACT_FROM    => 'lib/Service/WorkHours.pm',
	LICENSE          => 'artistic_2',
	PL_FILES         => {},
	MIN_PERL_VERSION => '5.014',
	EXE_FILES => [
		'scripts/workhoursd'
	],
	CONFIGURE_REQUIRES => {
		'ExtUtils::MakeMaker' => '0',
	},
	BUILD_REQUIRES => {
		'Test::More' => '0',
		'Test::CheckManifest' => '0.9',
		'Test::Exception' => '0',
		'IO::All' => '0',
		'Test::Deep' => '0'
	},
	PREREQ_PM => {
		'Cwd' => 0,
		'File::Basename' => 0,
		'File::Glob' => 0,
		'File::Spec' => 0,
		'Getopt::Long' => 0,
		'List::Util' => 0,
		'Net::DBus' => 0,
		'Pod::Usage' => 0,
		'Proc::Daemon' => 0,
		'YAML' => 0
	},
	dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
	clean => { FILES => 'Service-WorkHours-*' },
);


lib/Service/WorkHours/Config.pm
package Service::WorkHours::Config;

use 5.014;
use strict;
use warnings;
use Carp;

use File::Basename qw/dirname/;
use File::Glob ':bsd_glob';
use File::Spec::Functions qw/catfile file_name_is_absolute/;
use YAML;

=head1 NAME

Service::WorkHours::Config - Configuration class for workhoursd

=head1 SYNOPSIS

This class loads service specifications from a YAML formatted configuration file
and loads the corresponding units from a given systemd wrapper (see
L<Systemd::WorkHours::Systemd>).

	use Service::WorkHours::Config;
	use Service::WorkHours::Systemd;
	my $config = Service::WorkHours::Config->new(
		file => '/etc/workhoursd/main.yml',
		systemd => Service::WorkHours::Systemd->new
	);
	say keys %{$config->services};


=head1 METHODS

=cut

sub _str2timeofday {
	$_[0] =~ m/^(\d{1,2}):(\d{1,2})$/;
	my $t = $1 * 3600 + $2 * 60;

	die "Invalid time $t" if ($t < 0 || $t > (24*3600));
	return $t;
}

=head2 new(file => $path, systemd => $wrapper)

Loads a new configuration from the file specified by C<$path> and loads the
units corresponding to services from the given C<$wrapper>. Returns a new
instance representing this configuration.

=cut

sub new {
	my ($class, %opts) = @_;

	croak "file must be specified"
		unless $opts{file};

	croak "no systemd wrapper provided"
		unless $opts{systemd};

	# Default to includes enabled
	$opts{includes} //= 1;

	my $self = {
		services => {}
	};

	my @files = ref $opts{file} eq 'ARRAY' ?
	 			@{$opts{file}} : ($opts{file});
	my $loaded = 0;

	while (my $configfile = shift @files) {
		# Try to load config or warn
		my $config = eval { YAML::LoadFile($configfile) // {} };
		if ($@) {
			carp "cannot load config file $configfile: $@";
			next
		};

		my $dir = dirname($configfile);

		while (my ($k, $v) = each %{$config->{services} // {}}) {
			# Try to load the unit from systemd
			my $unit = defined $self->{services}->{$k} ?
						$self->{services}->{$k}->{unit} : # overriding, reuse unit
			 			$opts{systemd}->get_unit("$k.service");

			if ($unit) {
				my $svc = {
					name => $k,
					unit => $unit,
					startat => _str2timeofday($v->{start}),
					stopat => _str2timeofday($v->{stop}),
					ignorefailed => $v->{ignorefailed} // 0
				};

				# Check start and stop date
				if ($svc->{startat} >= $svc->{stopat}) {
					carp "'$k' is set to stop before it starts, ignoring.";
					next
				}

				# Unit has been found and is valid, register service
				$self->{services}->{$k} = $svc;
			} else {
				carp "'$k' has no matching service unit, ignoring.";
			}
		}

		# Process includes, unless disabled
		if ($opts{includes}) {
			if (ref $config->{include} eq 'ARRAY') {
				for my $inc (@{$config->{include}}) {
					# Each include can be a glob, which can itself be relative
					# to the current configuration file
					$inc = catfile($dir, $inc) unless
						file_name_is_absolute($inc);

					push @files, bsd_glob($inc);
				}
			}
		}

		$loaded++;
	}

	# Fail if no files could be loaded
	croak "could not load any config files" unless $loaded;

	bless $self, $class;
}

=head2 services()

Returns the services represented by this configuration instance.

=cut

sub services {
	my ($self) = @_;
	$self->{services}
}

=head1 AUTHOR

Vincent Tavernier, C<< <vince.tavernier at gmail.com> >>

=cut

1;


lib/Service/WorkHours/Systemd.pm
package Service::WorkHours::Systemd;

use 5.014;
use strict;
use warnings;
use Carp;

use Net::DBus;

=head1 NAME

Service::WorkHours::Systemd - Wrapper for access to systemd via DBus

=head1 SYNOPSIS

This class provides a wrapper for loading unit files from systemd through DBus.

	use Service::WorkHours::Systemd;
	my $systemd = Service::WorkHours::Systemd->new;
	say $systemd->get_unit("nginx.service")->ActiveState;

=head1 METHODS

=cut

=head2 new()

Initializes a new instance of the systemd DBus wrapper class.

=cut

sub new {
	my ($class, %opts) = @_;

	my $self = {
		bus => Net::DBus->system
	};

	$self->{systemd} = $self->{bus}->get_service("org.freedesktop.systemd1");
	$self->{manager} = $self->{systemd}->get_object("/org/freedesktop/systemd1");

	bless $self, $class;
}

=head2 get_unit($name)

Tries to load and retrieve the unit object associated with C<$name>. If the unit
cannot be loaded, C<undef> is returned.

=cut

sub get_unit {
	my ($self, $unitname) = @_;

	# Ask systemd to load the unit file
	my $unit = $self->{manager}->LoadUnit($unitname);

	if ($unit) {
		# Get the actual unit object
		$unit = $self->{systemd}->get_object($unit);

		# Print notification about loading
		my $path = $unit->FragmentPath || $unit->SourcePath;
		say STDERR "Loaded $unitname from $path";

		return $unit;
	} else {
		# No unit found
		return undef;
	}
}

=head1 AUTHOR

Vincent Tavernier, C<< <vince.tavernier at gmail.com> >>

=cut

1;


lib/Service/WorkHours.pm
package Service::WorkHours;

use 5.014;
use strict;
use warnings;

use Proc::Daemon;
use Cwd qw/abs_path/;
use Getopt::Long qw/GetOptionsFromArray/;
use List::Util qw/min/;

require Service::WorkHours::Config;
require Service::WorkHours::Systemd;

=head1 NAME

Service::WorkHours - Service for managing systemd services working hours

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

=head1 SYNOPSIS

Implementation of the Service::WorkHours daemon.

	workhoursd -d --config /etc/workhoursd/main.yml

=head1 OPTIONS

=over 8

=item B<--daemon>

Enable daemon mode. Use this when you want to manually start this as a daemon,
but don't use this option if you use systemd to start workhoursd as a service.

=item B<--config>

Path to the config file for service work hours specification.

=item B<--help>

Prints a brief help message and exits.

=item B<--man>

Prints the manual page and exits.

=back

=cut

sub _usage {
	pod2usage(@_, -input => abs_path(__FILE__));
}

=head1 SUBROUTINES

=head2 run(@args)

Implements the B<workhoursd> command.

=cut

sub run {
	shift;

	# Parse options
	my ($man, $help, $daemon, $configfile) = (0, 0, 0, '/etc/workhoursd/main.yml');
	shift; GetOptionsFromArray(\@_, 'help|?' => \$help,
									'man' => \$man,
									'daemon' => \$daemon,
									'config=s' => \$configfile);

	# Print help
	_usage(-exitval => 1) if $help;
	_usage(-exitval => 0, -verbose => 2) if $man;

	# Switch to daemon mode if necessary
	Proc::Daemon::Init if $daemon;

	# State variables
	my ($continue, $reload) = (1, 0);

	# Setup signal handlers
	$SIG{TERM} = sub { $continue = 0 };
	$SIG{HUP} = sub { $reload = 1 };
	$SIG{INT} = sub { $continue = 0 };

	# Allocate systemd wrapper
	my $systemd = Service::WorkHours::Systemd->new;

	CONFIGLOOP:
	while ($continue) {
		# Load config
		my $config = Service::WorkHours::Config->new(file => $configfile,
													 systemd => $systemd);

		# Config reloaded
		$reload = 0;

		RUNLOOP:
		while ($continue && !$reload) {
			my $nextevent = 24*3600 + 1;

			my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) =
				localtime(time);
			my $t = $sec + $min * 60 + $hour * 3600;

			while (my ($k, $svc) = each %{$config->services}) {
				my $unit = $svc->{unit};

				# Is the unit active?
				my $active = ($unit->ActiveState =~ m/^active|reloading|activating$/);

				# Should the unit be active?
				my $should = $t >= $svc->{startat} && $t < $svc->{stopat};

				if ($active && !$should) {
					say STDERR "$k: stopping service";
					$unit->Stop("fail")
				} elsif (!$active && $should) {
					if ($unit->ActiveState eq 'failed' && !$svc->{ignorefailed}) {
						say STDERR "$k: not starting service in failed mode";
					} else {
						say STDERR "$k: starting service";
						$unit->Start("fail");
					}
				} elsif ($active && $should) {
					say STDERR "$k: running as expected";
				} elsif (!$active && !$should) {
					my $t = $svc->{startat};
					say STDERR "$k: waiting for $t";
				}

				my ($timetostart, $timetoend) = map { my $a = $svc->{$_} - $t;
														$a <= 0 ? 24*3600+$a : $a } qw/startat stopat/;
				$nextevent = min($nextevent, $timetostart, $timetoend);
			}

			if ($nextevent == 24*3600+1) {
				sleep;
			} else {
				sleep $nextevent;
			}
		}
	}
}

=head1 AUTHOR

Vincent Tavernier, C<< <vince.tavernier at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Vincent Tavernier.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Service::WorkHours
